### PR TITLE
Add support for Mastercard 2-Series BIN

### PR DESCRIPTION
**Pull request by prettyboymp**
Introduce support for Mastercard's soon to be released 2-Series BIN.  See https://www.mastercard.us/en-us/issuers/get-support/2-series-bin-expansion.html.

The new BIN series introduces the range of 2221 - 2720 as valid numbers for Mastercard.

To test:
`npm run test-client client/lib/credit-card-details`

### CODE REVIEW

**Comment by prettyboymp** [comment]
cc @gziolo

**Comment by gziolo** [comment]
Looks good to me. I tested our existing form and it works properly. I couldn't find any existing example of valid card number in the range of 2221 - 2720. All online generators I checked have this range disabled. So I can't confirm it is going to work with our validation logic, but at least it is able to recognise card type properly.

**Comment by gziolo** [comment]
We are using https://www.npmjs.com/package/creditcards as a dependency. It looks like it hasn't been updated in the last 3 months. We are using version `1.1.1` at the moment, but there is already version `2.1.2` which depends on `creditcards-types` package that has already proper pattern provided for Mastercard: https://github.com/bendrucker/creditcards-types/blob/master/src/types.js#L30. We should also update this dependency to make sure it all works properly together.

**Comment by prettyboymp** [comment]
@gziolo I've updated the https://www.npmjs.com/package/creditcards dependency to 2.1.2.  I was hoping to just use the creditcards-types handling directly, but it doesn't work against the `622126-622925` range for discover cards.

**Comment by gziolo** [comment]
This PR needs to be rebased with the master branch and new shrinkwrap needs to be generated. It looks like there was some `package.json` change introduced recently.

I was trying to test it in the browser. When you leave `Expiration date` field empty and try to fill in other field then JavaScript errors when trying to perform validation:

screen shot 2016-10-03 at 09 50 23

File: client/lib/credit-card-details/validation.js [hunk 1]
```diff
@@ -54,6 +54,20 @@ function creditCardFieldRules() {
 	};
 }
 
+function parseExpiration(value) {
+	const [month, year] = value.split( '/' );
```
Comment (gziolo) [inline client/lib/credit-card-details/validation.js]: We put spacing almost everywhere in JS code, similar to PHP. See [Spacing docs](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#spacing).

So there are couple of places where you need to update code in this PR.

I also recommend installing git pre-commit hook. See [here](https://github.com/Automattic/wp-calypso/blob/56aa9fbc09166180ce15c9eeb6a3f56cf32f7e80/Makefile#L181) or simply run from the command line:
`make githooks-commit`

File: client/lib/credit-card-details/validation.js [hunk 2]
```diff
@@ -68,54 +82,30 @@ validators.required = {
...
-		expirationMonth: expirationMonth,
-		expirationYear: expirationYear,
-		cvc: cardDetails.cvv
-	} );
-}
-
-function creditCardValidator( ...validationProperties ) {
-	return {
-		isValid: function( value, cardDetails ) {
-			if ( ! value ) {
```
Comment (gziolo) [inline client/lib/credit-card-details/validation.js]: I think we are missing this check in all refactored validators.

**Comment by gziolo** [comment]
Thanks for taking care of the library version updated and doing very nice refactoring! 👍

File: client/lib/credit-card-details/validation.js (hunk 2 above)
Comment (prettyboymp) [inline client/lib/credit-card-details/validation.js]: Shouldn't the `required` rule that each field uses already handle that clause?

Comment (prettyboymp) [inline client/lib/credit-card-details/validation.js]: Nevermind, I see that the code loops through all rules for a field even if a previous rule failed.

Comment (gziolo) [inline client/lib/credit-card-details/validation.js]: At the moment we use `required` check only when submitting the form. In effect we shouldn't validated all fields when user starts the process and fills in only one field and want to go to another one. We should only validate on focus change those field that are not empty.

**Comment by prettyboymp** [comment]
This branch has been rebased, squashed, and updated with the requested changes.

**Comment by gziolo** [comment]
All good now! Thanks for all improvements :) You can 🚢

**Pull request merged**